py/dml/toplevel.py

```diff
@@ -140,11 +140,24 @@ def scan_statements(filename, site, stmts):
         elif s.kind == 'toplevel_if':
             [cond, tbranch, fbranch] = s.args
             scope = symtab.Symtab()
-            bsite = SimpleSite('<builtin>', dml_version=dml.globals.dml_version)
-            scope.add(ctree.ExpressionSymbol(
+            bsite = SimpleSite('<builtin>',
+                               dml_version=dml.globals.dml_version)
+            # HACK Add constants to scope typically defined by dml-builtins,
```

Comment:
Ew, but not much to do about this I assume?

Reply:
Well, it's probably possible to eliminate the hack, but it would require more intricate changes which I didn't want to waste my time looking into as the current implementation of experimental `#if`s is very unlikely to remain, anyway.

Reply:
i dont think these added constants have a use case. isn't it easier to read/remember/explain 'dml_1_2 is the only permitted symbol' instead of 'the only allowed symbols are dml_1_2 plus some random useless symbols'?

Reply:
`unmapped_offset` I can understand, but I certainly would not call `true` and `false` "random useless symbols" -- people will get understandably confused if they see an error message telling them that "false" is an unknown identifier. Adding `dml_1_4` is nice for symmetry if we already support `dml_1_2`.

Reply:
what's the use case for `#if (false)`?
dml_1_4 is meant for distinguishing between 1.4 and 1.5 in code shared between the two versions; this use case doesn't exist yet, we just reserved the ident.

Reply:
If nothing else, `#if (false)` is a nice way to comment out code; and for that reason alone people _will_ reach for it, in which case we don't want to say "unknown identifier: `false`", because that will only serve to confuse people. Just see the HSD-ES ticket.

Reply:
Oh right, sorry, you're on vacation; I keep forgetting this repo is public :)

Reply:
right, that's a fair use of `false`. however, what would truly reduce confusion is an error message that clarifies stuff, like 'only [false and] dml_1_2 allowed as conditions of #if blocks containing %s'. hard to write that short with a long list of idents.

Reply:
We don't write that down regardless. Besides, despite `false` and `true` being identifiers in DML 1.4 people typically view them as literals instead, so perhaps we don't need to explicitly say they may be used. I'm more than fine with scrapping `unmapped_offset` (I only added it because it was a `constant`, rather than a `param`) and while I won't protest to scrapping `dml_1_4` as well, doing so feels asymmetrical and weird.

Reply:
the report mentions 'if false' but that's likely just a minimal reproducer. the original use case was likely a param. I don't think their debug confusion would be any smaller if the behaviour further changed when moving from a false param to false. Just changes the reproducer.

hyrum requires you to write test cases for all allowed idents and those test cases will be weird. (dml_1_2 is naturally smoke tested because it has strong use cases).

the dml_1_2 hack is hideous and by nature asymmetrical; to me, extending it with other arbitrary idents seems to increase that asymmetry. the current pattern is that any idents absolutely needed for survival are included.

Reply:
that said, i'd say the change is neither significantly harmful nor useful, so i don't object to merge if you insist (but test is needed for all idents). just wanted to explain my reasoning.

Reply:
I insist at least for `true` and `false`. `dml_1_4` can be scrapped.